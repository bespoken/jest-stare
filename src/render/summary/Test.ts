import { Constants } from "../Constants";

/**
 * Create elements for a test
 * @export
 * @class Test
 */
export class Test {

    /**
     * Add test to a section of the presented table
     * @static
     * @param {IInnerTestResults} innerTestResult - see object description
     * @param {HTMLDivElement} element - base element to add to
     * @returns {HTMLDivElement} - populated element
     * @memberof Test
     */
    public static create(innerTestResult: jest.AssertionResult): HTMLElement {
        const containerDiv = document.createElement("div") as HTMLDivElement;

        const anchor = document.createElement("a") as HTMLAnchorElement;
        anchor.href = "#" + innerTestResult.title.replace(/\s+/g, "-").toLowerCase();

        const testName = document.createElement("span") as HTMLElement;
        testName.textContent = innerTestResult.title;

        anchor.appendChild(Test.getSimbolSpanFromStatus(innerTestResult.status));
        anchor.appendChild(testName);

        containerDiv.appendChild(anchor);

        return containerDiv;
    }

    /**
     * Generates a span with a symbol and appropiate color from the test status
     * @static
     * @param {status} string - the test status
     * @returns {HTMLSpanElement} - populated element
     * @memberof Test
     */
    private static  getSimbolSpanFromStatus(status: string) {
        const span = document.createElement("span") as HTMLSpanElement;
        span.classList.add("summary-test-label", "test");

        if (status === Constants.TEST_STATUS_PASS) {
            span.textContent = "âœ“";
            span.classList.add("pass");
        }

        if (status === Constants.TEST_STATUS_PEND) {
            span.textContent = "O";
            span.classList.add("pending");
        }

        if (status === Constants.TEST_STATUS_FAIL) {
            span.textContent = "X";
            span.classList.add("fail");
        }

        return span;
    }
}
