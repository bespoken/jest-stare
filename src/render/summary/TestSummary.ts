import { Constants } from "../Constants";

/**
 * Create test summary
 * @export
 * @class TestSummary
 */
export class TestSummary {
    /**
     * Ancestor title join character
     * @static
     * @memberof TestSummary
     */
    public static readonly JOIN_CHAR = ".";

    /**
     * Build table info for specific tests
     * @static
     * @returns {HTMLElement[]} - populated html elements
     * @memberof TestSummary
     */
    public static create(results: jest.AggregatedResult): HTMLElement[] {
        const elements: HTMLElement[] = [];

        const div = document.createElement("div") as HTMLDivElement;
        div.classList.add("my-3", "p-3", "bg-white", "rounded", "box-shadow", "summary");

        const h5 = document.createElement("h5") as HTMLHeadingElement;
        h5.classList.add("border-bottom", "pb-2", "display-5", "summary-title");
        h5.textContent = "Summary";

        div.appendChild(h5);
        div.id = "test-summary";
        elements.push(div);

        results.testResults.forEach((testResult, pathIndex) => {

            // NOTE(Kelosky): jest.AggregateResult has a testResults array
            // which contains a jest.TestResults array.  jest.TestResults array
            // is of type AssertionResults array.  however, it looks like they
            // somehow allow for the the field name to be assertionResults instead
            // of the documented interface testResults.  So, we'll cast to any, and attempt
            // access assertionResults if testsResults are missing
            if (testResult.testResults == null) {
                // tslint:disable-next-line:no-console
                console.error("Unexpected testResults field missing");
                if ((testResult as any).assertionResults != null) {
                    // tslint:disable-next-line:no-console
                    console.warn("Attempting to use assertionResults: results are unpredictable");
                    testResult.testResults = (testResult as any).assertionResults;
                }
            }

            const divMap: Map<string, HTMLElement> = new Map<string, HTMLElement>();
            const testCountMap: Map<string, any> = new Map<string, any>();
            const testCountArray: string[] = [];

            const testTitleDiv = document.createElement("div") as HTMLDivElement;
            testTitleDiv.classList.add("summary-test-suite");

            const testFileLink = document.createElement("a") as HTMLAnchorElement;
            const isFail = testResult.numFailingTests > 0;

            const testStatus = document.createElement("strong") as HTMLSpanElement;

            testStatus.classList.add("summary-test-label");
            if (!isFail) {
                testStatus.classList.add("pass");
                testStatus.textContent = "PASS";
            } else {
                testStatus.classList.add("fail");
                testStatus.textContent = "FAIL";
            }

            const testFileLine = document.createElement("strong") as HTMLSpanElement;
            testFileLine.classList.add("summary-test-label", "path");
            testFileLine.textContent = testResult.testFilePath;
            testFileLink.href = "#" + testResult.testFilePath;

            testFileLink.appendChild(testStatus);
            testFileLink.appendChild(testFileLine);

            testTitleDiv.appendChild(testFileLink);
            div.appendChild(testTitleDiv);

            testResult.testResults.forEach((test, testIndex) => {
                if (test.ancestorTitles.length > 0) {
                    test.ancestorTitles.forEach((title, index) => {

                        const titlesCopy = test.ancestorTitles.slice();
                        titlesCopy.splice(index + 1);
                        const key = titlesCopy.join(TestSummary.JOIN_CHAR);
                        if (!divMap.has(key)) {
                            const nestDiv = document.createElement("div") as HTMLDivElement;
                            nestDiv.classList.add("summary-ancestor-box");
                            nestDiv.id = key;

                            divMap.set(key, nestDiv);

                            if (index === 0) {
                                div.appendChild(nestDiv);
                            } else {
                                titlesCopy.pop();
                                const parentKey = titlesCopy.join(TestSummary.JOIN_CHAR);
                                divMap.get(parentKey).appendChild(nestDiv);

                                const href = "#" + key + TestSummary.JOIN_CHAR
                                    + pathIndex + TestSummary.JOIN_CHAR + testIndex
                                    + TestSummary.JOIN_CHAR + index;
                                testCountMap.set(key, {
                                    locale: test.ancestorTitles[0],
                                    title,
                                    href,
                                    passingTests: 0,
                                    pendingTests: 0,
                                    failingTests: 0
                                });
                                testCountArray.push(key);

                            }
                        }

                        if (index !== 0) {
                            const testCount = testCountMap.get(key);
                            if (test.status === Constants.TEST_STATUS_PASS) {
                                testCount.passingTests = testCount.passingTests + 1;
                            }

                            if (test.status === Constants.TEST_STATUS_PEND) {
                                testCount.pendingTests = testCount.pendingTests + 1;
                            }

                            if (test.status === Constants.TEST_STATUS_FAIL) {
                                testCount.failingTests = testCount.failingTests + 1;
                            }
                            testCountMap.set(key, testCount);
                        }
                    });
                }
            });

            const locales =  [...new Set(testCountArray.map((key) => testCountMap.get(key).locale))];

            testCountArray.forEach((key) => {
                const testCount = testCountMap.get(key);
                const testCountSpan = document.createElement("strong") as HTMLSpanElement;
                testCountSpan.classList.add("summary-test-count");

                const testTotal = testCount.passingTests + testCount.pendingTests + testCount.failingTests;
                testCount.testTotal = testTotal;
                const testCountText = "[" + testCount.passingTests + "/" + testTotal + "]";
                testCountSpan.textContent = testCountText;

                const containerDiv = divMap.get(key);
                const ancestorLink = document.createElement("a") as HTMLAnchorElement;
                ancestorLink.textContent = locales.length > 1 ? `(${testCount.locale}) ${testCount.title}`
                    : testCount.title;
                ancestorLink.href = testCount.href;
                ancestorLink.classList.add("summary-test-label", "path");

                containerDiv.appendChild(TestSummary.getSimbolSpanFromStatus(testCount));
                containerDiv.appendChild(ancestorLink);
                containerDiv.appendChild(testCountSpan);
            });

        });

        return elements;
    }

    /**
     * Generates a span with a symbol and appropiate color from the test status
     * @static
     * @param {status} string - the test status
     * @returns {HTMLSpanElement} - populated element
     * @memberof TestSummary
     */
    private static getSimbolSpanFromStatus({testTotal, passingTests, pendingTests, failingTests}) {
        const span = document.createElement("span") as HTMLSpanElement;
        span.classList.add("summary-test-label", "test", "smallLabel");

        if (failingTests > 0) {
            span.textContent = "X";
            span.classList.add("fail");
        } else if (pendingTests === testTotal) {
            span.textContent = "O";
            span.classList.add("pending");
        } else if (passingTests + pendingTests === testTotal) {
            span.textContent = "âœ“";
            span.classList.add("pass");
        }

        return span;
    }
}
