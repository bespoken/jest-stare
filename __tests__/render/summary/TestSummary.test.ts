jest.mock("../../../src/render/summary/Test");

import { TestSummary } from "../../../src/render/summary/TestSummary";
import { Test } from "../../../src/render/summary/Test";

const simplePassingTests: jest.AggregatedResult = require("../../__resources__/simplePassingTests.json");
const simpleFailingTests: jest.AggregatedResult = require("../../__resources__/simpleFailingTests.json");
const twoSideBySideFailDiffTests: jest.AggregatedResult = require("../../__resources__/twoSideBySideFailDiffTests.json");

// mock out the individual tests within describe blocks
(Test.create as any) = jest.fn( () => {
    const div = document.createElement("div") as HTMLDivElement;
    div.textContent = "dummy div";
    return div;
});

describe("TestSuite tests", () => {

    it("should create proper elements for simple passing tests", () => {
        expect(TestSummary.create(simplePassingTests)).toMatchSnapshot();
    });

    it("should create proper elements for simple failing tests", () => {
        expect(TestSummary.create(simpleFailingTests)).toMatchSnapshot();
    });

    it("should create proper elements for two diffs in html", () => {
        expect(TestSummary.create(twoSideBySideFailDiffTests)).toMatchSnapshot();
    });

    it("should create proper elements for large test > 500 tests", () => {
        expect(TestSummary.create(twoSideBySideFailDiffTests)).toMatchSnapshot();
    });
});
