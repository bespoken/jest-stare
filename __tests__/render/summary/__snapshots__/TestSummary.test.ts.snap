// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestSuite tests should create proper elements for large test > 500 tests 1`] = `
Array [
  <div
    class="my-3 p-3 bg-white rounded box-shadow summary"
    id="test-summary"
  >
    <h5
      class="border-bottom pb-2 display-5 summary-title"
    >
      Summary
    </h5>
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\example\\\\FailingTests.test.ts"
      >
        <strong
          class="summary-test-label fail"
        >
          FAIL
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\example\\FailingTests.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Failing Tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\render\\\\Render.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\render\\Render.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Render tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\processor\\\\Processor.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\processor\\Processor.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Processor tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\utils\\\\IO.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\utils\\IO.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="IO tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\utils\\\\Logger.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\utils\\Logger.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Logger tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\reporter\\\\Reporter.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\reporter\\Reporter.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Reporter tests"
    />
  </div>,
]
`;

exports[`TestSuite tests should create proper elements for simple failing tests 1`] = `
Array [
  <div
    class="my-3 p-3 bg-white rounded box-shadow summary"
    id="test-summary"
  >
    <h5
      class="border-bottom pb-2 display-5 summary-title"
    >
      Summary
    </h5>
    <div
      class="summary-test-suite"
    >
      <a
        href="#Work\\\\Dev\\\\node\\\\jest-stare\\\\__tests__\\\\utils\\\\IO.test.ts"
      >
        <strong
          class="summary-test-label fail"
        >
          FAIL
        </strong>
        <strong
          class="summary-test-label path"
        >
          Work\\Dev\\node\\jest-stare\\__tests__\\utils\\IO.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="IO tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#Work\\\\Dev\\\\node\\\\jest-stare\\\\__tests__\\\\reporter\\\\Reporter.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          Work\\Dev\\node\\jest-stare\\__tests__\\reporter\\Reporter.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Reporter tests"
    />
  </div>,
]
`;

exports[`TestSuite tests should create proper elements for simple passing tests 1`] = `
Array [
  <div
    class="my-3 p-3 bg-white rounded box-shadow summary"
    id="test-summary"
  >
    <h5
      class="border-bottom pb-2 display-5 summary-title"
    >
      Summary
    </h5>
    <div
      class="summary-test-suite"
    >
      <a
        href="#Work\\\\Dev\\\\node\\\\jest-stare\\\\__tests__\\\\utils\\\\IO.test.ts"
      >
        <strong
          class="summary-test-label fail"
        >
          FAIL
        </strong>
        <strong
          class="summary-test-label path"
        >
          Work\\Dev\\node\\jest-stare\\__tests__\\utils\\IO.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="IO tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#Work\\\\Dev\\\\node\\\\jest-stare\\\\__tests__\\\\reporter\\\\Reporter.test.ts"
      >
        <strong
          class="summary-test-label fail"
        >
          FAIL
        </strong>
        <strong
          class="summary-test-label path"
        >
          Work\\Dev\\node\\jest-stare\\__tests__\\reporter\\Reporter.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Reporter tests"
    />
  </div>,
]
`;

exports[`TestSuite tests should create proper elements for two diffs in html 1`] = `
Array [
  <div
    class="my-3 p-3 bg-white rounded box-shadow summary"
    id="test-summary"
  >
    <h5
      class="border-bottom pb-2 display-5 summary-title"
    >
      Summary
    </h5>
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\example\\\\FailingTests.test.ts"
      >
        <strong
          class="summary-test-label fail"
        >
          FAIL
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\example\\FailingTests.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Failing Tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\render\\\\Render.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\render\\Render.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Render tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\processor\\\\Processor.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\processor\\Processor.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Processor tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\utils\\\\IO.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\utils\\IO.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="IO tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\utils\\\\Logger.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\utils\\Logger.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Logger tests"
    />
    <div
      class="summary-test-suite"
    >
      <a
        href="#jest-stare\\\\__tests__\\\\src\\\\reporter\\\\Reporter.test.ts"
      >
        <strong
          class="summary-test-label pass"
        >
          PASS
        </strong>
        <strong
          class="summary-test-label path"
        >
          jest-stare\\__tests__\\src\\reporter\\Reporter.test.ts
        </strong>
      </a>
    </div>
    <div
      class="summary-ancestor-box"
      id="Reporter tests"
    />
  </div>,
]
`;
